fontra2fontir/src/fontra.rs

```diff
@@ -3,21 +3,105 @@
 //! See <https://github.com/googlefonts/fontra/blob/main/src/fontra/core/classes.py>
 
 use std::{
-    collections::HashMap,
+    collections::{BTreeMap, HashMap},
     fs,
     path::{Path, PathBuf},
 };
 
 use fontdrasil::types::GlyphName;
 use fontir::error::Error;
-use percent_encoding::{utf8_percent_encode, NON_ALPHANUMERIC};
 use serde::Deserialize;
 use write_fonts::types::Tag;
 
+const SEPARATOR_CHAR: char = '^';
+
+fn is_reserved_char(c: char) -> bool {
+    let c = c as u32;
+    match c {
+        _ if c < 32 => true,
+        _ if c == 0x7F => true,
+        _ if c == SEPARATOR_CHAR as u32 => true,
+        _ => false,
+    }
+}
+
+fn is_reserved_filename(name: &str) -> bool {
+    matches!(
+        name.to_ascii_uppercase().as_str(),
+        "CON"
+            | "PRN"
+            | "AUX"
+            | "CLOCK$"
+            | "NUL"
+            | "COM1"
+            | "LPT1"
+            | "LPT2"
+            | "LPT3"
+            | "COM2"
+            | "COM3"
+            | "COM4"
+    )
+}
+
+const BASE_32_CHARS: [char; 32] = [
+    '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I',
+    'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V',
+];
+
+/// Matches <https://github.com/googlefonts/fontra/blob/15bc0b8401054390484cfb86d509d633d29657a1/src/fontra/backends/filenames.py#L40-L64>
+fn string_to_filename(string: &str, suffix: &str) -> String {
+    let mut code_digits: Vec<_> = (0..string.len())
+        .step_by(5)
+        .map(|i| {
+            let mut digit = 0;
+            let mut bit = 1;
+            for c in string[i..(i + 5).min(string.len())].chars() {
+                if c.is_ascii_uppercase() {
+                    digit |= bit
+                }
+                bit <<= 1;
+            }
+            digit
+        })
+        .collect();
+    while let Some(0) = code_digits.last() {
+        code_digits.pop();
+    }
+
+    let mut filename = Vec::new();
```

Comment:
It might be easier to just declare this as a `String`? You still get `push` which takes a `char`, but you also get `push_str` to concatenate, which would simplify the code below.